File: mcserver/curseforge.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple
from urllib.error import HTTPError

from .errors import InvalidApiKeyError, MissingApiKeyError, UserFacingError
from .config import AppConfig
from .http_client import http_get_json


@dataclass(frozen=True)
class ModFile:
    id: int
    display_name: str
    file_date: str
    is_server_pack: bool
    server_pack_file_id: Optional[int]
    download_url: Optional[str]


class CurseForgeClient:
    BASE_URL = "https://api.curseforge.com"

    def __init__(self, api_key: Optional[str] = None):
        cfg = AppConfig.load()
        self.api_key = api_key or cfg.curseforge_api_key
        if not self.api_key:
            raise MissingApiKeyError(
                "Missing CurseForge API key. Run: mcserver config set-api-key"
            )

    def _wrap_http_errors(self, fn, *args, **kwargs):
        try:
            return fn(*args, **kwargs)
        except HTTPError as e:
            if e.code == 403:
                raise InvalidApiKeyError(
                    "CurseForge API returned 403 Forbidden (API key invalid). "
                    "Update it with: mcserver config set-api-key"
                )
            raise UserFacingError(
                f"CurseForge API request failed: HTTP {e.code} {e.reason}"
            )

    def _headers(self) -> Dict[str, str]:
        return {"Accept": "application/json", "x-api-key": self.api_key}

    def search_modpacks(
        self,
        *,
        query: str,
        game_version: Optional[str] = None,
        index: int = 0,
        page_size: int = 10,
        sort_field: int = 2,
        sort_order: str = "desc",
    ) -> List[Dict[str, Any]]:
        params: Dict[str, Any] = {
            "gameId": 432,
            "classId": 4471,
            "index": index,
            "pageSize": page_size,
            "searchFilter": query,
            "sortField": sort_field,
            "sortOrder": sort_order,
        }
        if game_version:
            params["gameVersion"] = game_version
        payload = self._wrap_http_errors(
            http_get_json,
            f"{self.BASE_URL}/v1/mods/search",
            headers=self._headers(),
            params=params,
        )
        return payload.get("data", [])

    def resolve_pack_id_from_url(self, url: str) -> int:
        match = re.search(r"/modpacks/([^/?#]+)", url)
        if not match:
            raise UserFacingError(
                "Invalid CurseForge modpack URL (expected /minecraft/modpacks/<slug>)."
            )
        slug = match.group(1)
        query = slug.replace("-", " ")
        results = self.search_modpacks(query=query, page_size=5)
        if not results:
            raise UserFacingError("No modpack found for the given URL.")
        return int(results[0]["id"])

    def list_files(self, pack_id: int) -> List[ModFile]:
        payload = self._wrap_http_errors(
            http_get_json,
            f"{self.BASE_URL}/v1/mods/{pack_id}/files",
            headers=self._headers(),
        )
        files: List[ModFile] = []
        for item in payload.get("data", []):
            files.append(
                ModFile(
                    id=int(item.get("id")),
                    display_name=str(item.get("displayName", "")),
                    file_date=str(item.get("fileDate", "")),
                    is_server_pack=bool(item.get("isServerPack", False)),
                    server_pack_file_id=item.get("serverPackFileId"),
                    download_url=item.get("downloadUrl"),
                )
            )
        return files

    def get_download_url(self, pack_id: int, file_id: int) -> str:
        payload = self._wrap_http_errors(
            http_get_json,
            f"{self.BASE_URL}/v1/mods/{pack_id}/files/{file_id}/download-url",
            headers=self._headers(),
        )
        data = payload.get("data")
        if not data:
            raise UserFacingError("Could not resolve download URL from CurseForge API.")
        return str(data)

    def choose_latest_server_pack(self, pack_id: int) -> Tuple[int, str, str]:
        """Returns (server_pack_file_id, display_name, file_date)."""
        files = self.list_files(pack_id)
        if not files:
            raise UserFacingError("No files found for this modpack.")

        # Prefer explicit server packs; otherwise use newest file that points to serverPackFileId.
        server_candidates = [f for f in files if f.is_server_pack]
        if server_candidates:
            server_candidates.sort(key=lambda f: f.file_date, reverse=True)
            f = server_candidates[0]
            return f.id, f.display_name, f.file_date

        # Fallback: newest file overall that has serverPackFileId
        files.sort(key=lambda f: f.file_date, reverse=True)
        newest = files[0]
        if newest.server_pack_file_id:
            return (
                int(newest.server_pack_file_id),
                newest.display_name,
                newest.file_date,
            )

        raise UserFacingError("No server pack found for this modpack.")

    def resolve_server_pack_download(
        self, pack_id: int, *, file_id: Optional[int] = None
    ) -> Tuple[str, int, str]:
        """Returns (download_url, server_pack_file_id, display_name)."""
        if file_id is None:
            server_file_id, display_name, _file_date = self.choose_latest_server_pack(
                pack_id
            )
        else:
            # User provided a file id. It might be a server pack or a normal file.
            # We resolve it by scanning for an exact id match; if it's not a server pack
            # but has serverPackFileId, use that.
            files = self.list_files(pack_id)
            match = next((f for f in files if f.id == file_id), None)
            if not match:
                raise UserFacingError(
                    f"File id {file_id} not found for pack {pack_id}."
                )
            display_name = match.display_name
            if match.is_server_pack:
                server_file_id = match.id
            elif match.server_pack_file_id:
                server_file_id = int(match.server_pack_file_id)
            else:
                raise UserFacingError(
                    "Selected file does not have an associated server pack."
                )

        url = self.get_download_url(pack_id, server_file_id)
        return url, int(server_file_id), str(display_name)

File: internal/curseforge/curseforge.go
package curseforge

import (
	"fmt"
	"regexp"
	"sort"
	"strconv"
	"strings"

	mcerrors "github.com/ericsson/curseforge-server-installer/internal/errors"
	mchttp "github.com/ericsson/curseforge-server-installer/internal/http"
)

const baseURL = "https://api.curseforge.com"

// ModFile represents a file entry from CurseForge.
type ModFile struct {
	ID                int
	DisplayName       string
	FileDate          string
	IsServerPack      bool
	ServerPackFileID  *int
	DownloadURL       string
}

// Client interacts with the CurseForge API.
type Client struct {
	apiKey string
}

// NewClient creates a CurseForge client with the given API key.
func NewClient(apiKey string) (*Client, error) {
	if apiKey == "" {
		return nil, mcerrors.NewMissingApiKeyError("Missing CurseForge API key. Run: cfs config set-api-key")
	}
	return &Client{apiKey: apiKey}, nil
}

func (c *Client) headers() map[string]string {
	return map[string]string{
		"Accept":    "application/json",
		"x-api-key": c.apiKey,
	}
}

func (c *Client) wrapHTTPError(err error) error {
	if httpErr, ok := err.(*mchttp.HTTPError); ok {
		if httpErr.StatusCode == 403 {
			return mcerrors.NewInvalidApiKeyError(
				"CurseForge API returned 403 Forbidden (API key invalid). Update it with: cfs config set-api-key")
		}
		return mcerrors.NewUserFacingErrorf("CurseForge API request failed: HTTP %d", httpErr.StatusCode)
	}
	return err
}

// SearchModpacks searches for modpacks matching the query.
func (c *Client) SearchModpacks(query string, gameVersion string, pageSize int) ([]map[string]any, error) {
	params := map[string]string{
		"gameId":       "432",
		"classId":      "4471",
		"index":        "0",
		"pageSize":     strconv.Itoa(pageSize),
		"searchFilter": query,
		"sortField":    "2",
		"sortOrder":    "desc",
	}
	if gameVersion != "" {
		params["gameVersion"] = gameVersion
	}
	payload, err := mchttp.GetJSON(baseURL+"/v1/mods/search", c.headers(), params)
	if err != nil {
		return nil, c.wrapHTTPError(err)
	}
	data, _ := payload["data"].([]any)
	results := make([]map[string]any, 0, len(data))
	for _, item := range data {
		if m, ok := item.(map[string]any); ok {
			results = append(results, m)
		}
	}
	return results, nil
}

// ResolvePackIDFromURL extracts a pack ID from a CurseForge modpack URL.
func (c *Client) ResolvePackIDFromURL(url string) (int, error) {
	re := regexp.MustCompile(`/modpacks/([^/?#]+)`)
	match := re.FindStringSubmatch(url)
	if match == nil {
		return 0, mcerrors.NewUserFacingError("Invalid CurseForge modpack URL (expected /minecraft/modpacks/<slug>).")
	}
	slug := match[1]
	query := strings.ReplaceAll(slug, "-", " ")
	results, err := c.SearchModpacks(query, "", 5)
	if err != nil {
		return 0, err
	}
	if len(results) == 0 {
		return 0, mcerrors.NewUserFacingError("No modpack found for the given URL.")
	}
	id, _ := results[0]["id"].(float64)
	return int(id), nil
}

// ListFiles returns all files for a modpack.
func (c *Client) ListFiles(packID int) ([]ModFile, error) {
	payload, err := mchttp.GetJSON(
		fmt.Sprintf("%s/v1/mods/%d/files", baseURL, packID),
		c.headers(), nil,
	)
	if err != nil {
		return nil, c.wrapHTTPError(err)
	}
	data, _ := payload["data"].([]any)
	files := make([]ModFile, 0, len(data))
	for _, item := range data {
		m, ok := item.(map[string]any)
		if !ok {
			continue
		}
		f := ModFile{
			ID:           intFromAny(m["id"]),
			DisplayName:  strFromAny(m["displayName"]),
			FileDate:     strFromAny(m["fileDate"]),
			IsServerPack: boolFromAny(m["isServerPack"]),
			DownloadURL:  strFromAny(m["downloadUrl"]),
		}
		if spfID := m["serverPackFileId"]; spfID != nil {
			v := intFromAny(spfID)
			if v != 0 {
				f.ServerPackFileID = &v
			}
		}
		files = append(files, f)
	}
	return files, nil
}

// GetDownloadURL returns the download URL for a specific file.
func (c *Client) GetDownloadURL(packID, fileID int) (string, error) {
	payload, err := mchttp.GetJSON(
		fmt.Sprintf("%s/v1/mods/%d/files/%d/download-url", baseURL, packID, fileID),
		c.headers(), nil,
	)
	if err != nil {
		return "", c.wrapHTTPError(err)
	}
	data, ok := payload["data"].(string)
	if !ok || data == "" {
		return "", mcerrors.NewUserFacingError("Could not resolve download URL from CurseForge API.")
	}
	return data, nil
}

// ChooseLatestServerPack returns (serverPackFileID, displayName, fileDate).
func (c *Client) ChooseLatestServerPack(packID int) (int, string, string, error) {
	files, err := c.ListFiles(packID)
	if err != nil {
		return 0, "", "", err
	}
	if len(files) == 0 {
		return 0, "", "", mcerrors.NewUserFacingError("No files found for this modpack.")
	}

	// Prefer explicit server packs
	var serverCandidates []ModFile
	for _, f := range files {
		if f.IsServerPack {
			serverCandidates = append(serverCandidates, f)
		}
	}
	if len(serverCandidates) > 0 {
		sort.Slice(serverCandidates, func(i, j int) bool {
			return serverCandidates[i].FileDate > serverCandidates[j].FileDate
		})
		f := serverCandidates[0]
		return f.ID, f.DisplayName, f.FileDate, nil
	}

	// Fallback: newest file with serverPackFileId
	sort.Slice(files, func(i, j int) bool {
		return files[i].FileDate > files[j].FileDate
	})
	newest := files[0]
	if newest.ServerPackFileID != nil {
		return *newest.ServerPackFileID, newest.DisplayName, newest.FileDate, nil
	}
	return 0, "", "", mcerrors.NewUserFacingError("No server pack found for this modpack.")
}

// ResolveServerPackDownload returns (downloadURL, serverPackFileID, displayName).
func (c *Client) ResolveServerPackDownload(packID int, fileID *int) (string, int, string, error) {
	var serverFileID int
	var displayName string

	if fileID == nil {
		var err error
		serverFileID, displayName, _, err = c.ChooseLatestServerPack(packID)
		if err != nil {
			return "", 0, "", err
		}
	} else {
		files, err := c.ListFiles(packID)
		if err != nil {
			return "", 0, "", err
		}
		var matched *ModFile
		for i := range files {
			if files[i].ID == *fileID {
				matched = &files[i]
				break
			}
		}
		if matched == nil {
			return "", 0, "", mcerrors.NewUserFacingErrorf("File id %d not found for pack %d.", *fileID, packID)
		}
		displayName = matched.DisplayName
		if matched.IsServerPack {
			serverFileID = matched.ID
		} else if matched.ServerPackFileID != nil {
			serverFileID = *matched.ServerPackFileID
		} else {
			return "", 0, "", mcerrors.NewUserFacingError("Selected file does not have an associated server pack.")
		}
	}

	url, err := c.GetDownloadURL(packID, serverFileID)
	if err != nil {
		return "", 0, "", err
	}
	return url, serverFileID, displayName, nil
}

func intFromAny(v any) int {
	switch n := v.(type) {
	case float64:
		return int(n)
	case int:
		return n
	case string:
		i, _ := strconv.Atoi(n)
		return i
	}
	return 0
}

func strFromAny(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func boolFromAny(v any) bool {
	if b, ok := v.(bool); ok {
		return b
	}
	return false
}
